### ISSUE TITLE
`par_sort_unstable` is relatively slow

### ISSUE THREAD

**Issue by RagnarGrootKoerkamp** [vigna/sux-rs#29]
https://github.com/vigna/sux-rs/blob/e081b427f99eeb5ce127dc74c5c03cd0862912c2/src/utils/sig_store.rs#L253

`Rayon::par_sort_unstable` does a comparison based merge sort. In my experiments the radix sort from the `rdst` crate is quite a bit faster on 1e9 keys (with `6` threads):
- `rdst`: `5.58s`
- `par_sort_unstable`: `9.39s`

You could also consider splitting the `SigStore` and `ChunkIterator`. For my usage I'd prefer to only have the `SigStore` functionality of partitioning into a file per bucket and iterating over those unsorted and unchunked, so that I can do this myself and chunk in the sizes I need.

**Comment by vigna** [comment]
Very slow. In Java I have implemented an optimized parallel radix sort (Mcllroy's) and use just that. Region sort (Obeya's) would be better but I don't think there's a Rust implementation.

I use for other purposes voracious_sort, which is amazing but does not work with [u64;2]. If rdst can support that type, absolutely, it's the way to go.

Note that the current development version of SigStore (in the result branch) does not check for duplicates, so it does not contain any sort. That's better left to the clients, for a number of reasons, but mostly because it makes a higher degree of concurrency possible.

**Comment by vigna** [comment]
Oh I see it does! Thanks for the tip, this is really great.

**Comment by vigna** [comment]
Actually, you need to implement RadixKey to sort on a key that's just part of the elements to sort, so I'd need to create a Sig structure. I asked the author if he could implement sort_by_key, which would be immensely useful and probably almost trivial to do. Then I'll switch, either way.

**Comment by RagnarGrootKoerkamp** [comment]
Yes `sort_by_key` would be nice indeed! (https://github.com/Nessex/rdst/issues/8, for reference, and as example my [implementation of `RadixKey`](https://github.com/RagnarGrootKoerkamp/PTRHash/blob/master/src/hash.rs#L18))

But yes agreed that this is better left for clients.

I've also tried `voracious` but `rdst` was faster. (Don't have the timings anymore.)
And note that `rdst` does implement Regions sort: https://github.com/Nessex/rdst/blob/8b7647956454055501f3e4c5882dd584c1830513/src/sorts/regions_sort.rs
(but it depends on the `Tuner` on whether it's actually used or not.)

**Comment by vigna** [comment]
I'm sold :). I was writing a RadixKey for SigSort but if I can get by_key I prefer. It's on hold.

**Comment by vigna** [comment]
Well, it turns out that it is relatively slow, but for the time being I cannot replace it...

https://github.com/Nessex/rdst/issues/10

**Comment by vigna** [comment]
Ok, implemented in result. Soon merged into main.
